File: problems/0020-valid-parentheses/solution.js
/**
 * @problemNumber: 20
 * @problemName: Valid Parentheses
 * @difficulty: Easy
 * @topics: String, Stack
 * @companies: Amazon, Facebook, Google
 */

/**
 * Approach: Stack
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * @param {string} s
 * @return {boolean}
 */
function isValid(s) {
    const stack = [];
    const map = {
        ')': '(',
        '}': '{',
        ']': '['
    };

    for (let char of s) {
        if (char === '(' || char === '{' || char === '[') {
            stack.push(char);
        } else {
            if (stack.length === 0 || stack.pop() !== map[char]) {
                return false;
            }
        }
    }

    return stack.length === 0;
}

// Test cases
console.log(isValid("()"));     // true
console.log(isValid("()[]{}")); // true
console.log(isValid("(]"));     // false

module.exports = { isValid };

File: problems/0020-valid-parentheses/solution.py
"""
Problem Number: 20
Problem Name: Valid Parentheses
Difficulty: Easy
Topics: String, Stack
Companies: Amazon, Facebook, Google
"""

class Solution:
    def isValid(self, s: str) -> bool:
        """
        Approach: Stack
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        stack = []
        mapping = {")": "(", "}": "{", "]": "["}
        
        for char in s:
            if char in mapping:
                top_element = stack.pop() if stack else '#'
                if mapping[char] != top_element:
                    return False
            else:
                stack.append(char)
        
        return not stack

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.isValid("()"))     # True
    print(solution.isValid("()[]{}")) # True
    print(solution.isValid("(]"))     # False
